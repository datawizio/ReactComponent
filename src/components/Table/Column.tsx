import * as React from "react";

import clsx from "clsx";
import { useState, useCallback, useMemo } from "react";

import { IColumn } from "./types";
import { PropsWithChildren } from "react";

export interface ColumnProps
  extends PropsWithChildren<any>,
    React.HTMLAttributes<any> {
  size: string;
  level: number;
  model: IColumn;
  multipleSorting: boolean;
}

// TODO убрать хардкод
function getColumnTopPos(level: number, size: string) {
  const columnHeightMap = {
    "small": 39,
    "middle": 47,
    "large": 55
  };

  return columnHeightMap[size] * (level - 1);
}

const Column: React.FC<ColumnProps> = props => {
  const {
    model,
    onClick,
    style,
    level,
    size,
    children,
    multipleSorting,
    ...restProps
  } = props;
  const [lastWidth, setLastWidth] = useState<number>(0);

  const onMouseDownHandler = useCallback(event => {
    setLastWidth(event.target.offsetWidth);
  }, []);

  const onClickHandler = useCallback(
    event => {
      const currentWidth = event.target.offsetWidth;
      lastWidth === currentWidth && onClick(event);
    },
    [lastWidth, onClick]
  );

  const styles = useMemo(
    () => ({
      ...style,
      top: getColumnTopPos(level, size),
      width: model.width || style.width
    }),
    [style, model.width, level, size]
  );

  const className = useMemo(() => {
    return clsx(
      "dw-table__column",
      {
        "dw-table__column--resizable": model.resizable,
        "dw-table__column--fixed": Boolean(model.fixed),

        "dw-table__column--fixed-left": model.fixed === "left",
        "dw-table__column--fixed-right": model.fixed === "right"
      },
      restProps.className
    );
  }, [model.resizable, model.fixed, restProps.className]);

  const sortSeq = useMemo(() => {
    return typeof model.sorter === "object" && model.sorter.multiple;
  }, [model.sorter]);

  return (
    <th
      {...restProps}
      style={styles}
      className={className}
      title={String(model.title)}
      onClick={onClickHandler}
      onMouseDown={onMouseDownHandler}
    >
      <div className="dw-table__column-content">
        {children}
        {sortSeq > 0 && (
          <span className="dw-table__column-sort-sequence">{sortSeq}</span>
        )}
      </div>
    </th>
  );
};

export default Column;
