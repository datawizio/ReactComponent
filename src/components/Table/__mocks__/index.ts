import { Action, BodyCellType, IColumn, TableState } from "../types";
import { withKnobs, text, boolean, number } from "@storybook/addon-knobs";
import { genColumns, genDataSource } from "../../../utils/data/dataGenerators";

const key = Math.random().toString(36).substring(7);

// defineCellType
export const validColumn: IColumn = {
  dtype: "dtype",
  dataIndex: "dataIndex",
  resizable: false
};

export const invalidColumn: IColumn = {
  dataIndex: "dataIndex",
  resizable: false
};

export const cell: BodyCellType = {
  dtype: "dtype",
  key: "key"
};

// filterByColumns
export const columnsMap: TableState["columnsMap"] = {
  col1: {
    dtype: "dtype",
    dataIndex: "1",
    resizable: true
  },
  col2: {
    dtype: "dtype2",
    dataIndex: "2",
    resizable: true
  },
  col3: {
    dtype: "dtype3",
    dataIndex: "3",
    resizable: true
  },
  col4: {
    dtype: "dtype4",
    dataIndex: "4",
    resizable: true
  }
};

export const obj: { [columnKey: string]: any } = {
  col1: {},
  col2: "any",
  col5: "string",
  col6: "number"
};

export const expectedOutput_filterByColumns = {
  col1: {},
  col2: "any"
};

// swapColumns
const generateCollumn = (isValid: boolean = true): IColumn => {
  const column = isValid ? { ...validColumn } : { ...invalidColumn };
  return {
    key,
    children: [
      {
        key,
        ...validColumn
      }
    ],
    ...column
  };
};

export const columnsMock: Array<IColumn> = [
  ...[...new Array(3)].map(() => generateCollumn(true)),
  ...[...new Array(2)].map(() => generateCollumn(false))
];

//Column

export const model = {
  key: "key",
  resizable: false,
  fixed: true,
  title: "title"
};

//Table

const columnsCount = number("columns count", 5);
const dataCount = number("data count", 5);
const treeColumns = boolean("treeColumns", true);
const treeData = boolean("treeData", true);
const toolBar = boolean("toolbar", true);
export const sortable = boolean("sortable", true);

export const columns = genColumns(columnsCount, treeColumns);

export let dataSource = genDataSource(dataCount, columns, ["string"], treeData);

if (columns[0]) {
  columns[0].fixed = "left";
  //
  columns[0].filters = [
    // @ts-ignore
    ...new Set(dataSource.map(item => item[columns[0].dataIndex]))
  ].map(item => ({
    text: item,
    value: item
  }));
}

// Handlers

export const dTypeConfig = {
  boolean: {},
  number: {},
  string: {}
};

// Table reducer

export const updateDataSourceAC = (payload): Action => ({
  type: "updateDataSource",
  payload: payload
});

export const updateColumnsAC = (payload): Action => ({
  type: "updateColumns",
  payload: payload || []
});

export const visibleColumnsKeysAC = (payload): Action => ({
  type: "visibleColumnsKeys",
  payload
});

export const paginateAC = (payload): Action => ({
  type: "paginate",
  payload
});

export const resetPaginationAC = (): Action => ({
  type: "resetPagination"
});

export const searchAC = (payload): Action => ({
  type: "search",
  payload
});

export const sorthAC = (payload): Action => ({
  type: "sort",
  payload: [].concat(payload)
});

export const filterAC = (payload): Action => ({
  type: "filter",
  payload
});

export const addLoadingRowAC = (payload): Action => ({
  type: "addLoadingRow",
  payload
});

export const setRowChildrenAC = (payload): Action => ({
  type: "setRowChildren",
  payload
});

export const swapColumnsAC = (payload): Action => ({
  type: "swapColumns",
  payload
});

export const expandRowAC = (payload): Action => ({
  type: "expandRow",
  payload
});

export const collapseRowAC = (payload): Action => ({
  type: "collapseRow",
  payload
});

export const loadingAC = (payload): Action => ({
  type: "loading",
  payload
});

export const recoveryStateAC = (payload): Action => ({
  type: "recoveryState",
  payload
});

export const updateRowAC = (payload): Action => ({
  type: "updateRow",
  payload
});

export const updateAC = (payload): Action => ({
  type: "update",
  payload
});
