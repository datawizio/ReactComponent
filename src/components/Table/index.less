@import "~antd/es/table/style/index";
@import "~antd/es/empty/style/index";
@import "~antd/es/pagination/style/index";

@import "../Select/index";
@import "../Loader/index";
@import "../../styles/variables";

.dw-table {
  &__wrapper {
    overflow: auto;
  }

  &__toolbar {
    padding: 0 5px;
    display: flow-root;
    position: relative;

    .table-toolbar--left {
      float: left;
      margin-right: 3px;
    }

    .table-toolbar--right {
      float: right;
      margin-right: 3px;
    }
  }

  .dw-table__column {
    width: 300px;
    transition: none !important;

    top: 0;
    z-index: 5;
    overflow: hidden;
    position: sticky !important;

    &--resizable {
      resize: horizontal;
    }

    &--drop-hover {
      background-color: tint(@primary-color, 80%) !important;
    }

    &--fixed {
      z-index: 10;

      &--left {
        left: 0;
      }

      &--right {
        right: 0;
      }
    }

    &-content {
      display: flex;
      flex-direction: row;
      align-items: center;
    }

    &-sort-sequence {
      color: @primary-color;
    }

    .ant-table-filter-trigger-container {
      background: transparent !important;
      right: 10px;
    }

    .ant-table-filter-column-title {
      padding: 8px !important;
    }
  }

  &__cell-data {
    display: flex;
    align-items: center;
  }

  table {
    table-layout: fixed;
    width: max-content !important;
  }

  td {
    .row-prefix {
      float: left;
      margin-right: 8px;
    }

    .ant-table-row-expand-icon {
      float: none;
    }

    position: relative;

    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  &--loading {
    table {
      .ant-empty {
        display: none;
      }
    }
  }

  &--empty {
    table {
      width: 100% !important;
      height: 100% !important;

      thead {
        display: none;
      }

      td {
        left: 50%;
        border: none !important;
        position: absolute !important;
      }
    }
  }

  .ant-table-bordered {
    .ant-table-cell {
      border-right: 1px solid #f0f0f0;
    }
  }

  .ant-table {
    .set-position-header(10, 55);

    &-small {
      .set-position-header(10, 39);
    }

    &-middle {
      .set-position-header(10, 47);
    }
  }
}

.set-columns-positions(@n, @size, @i: 2) when (@i =< @n) {
  tr:nth-child(@{i}) {
    .dw-table__column {
      top: unit(@size * (@i - 1), px);
    }
  }

  .set-columns-positions(@n, @size, (@i + 1));
}

.set-position-header(@n, @size) {
  thead {
    .set-columns-positions(@n, @size);
  }
}
