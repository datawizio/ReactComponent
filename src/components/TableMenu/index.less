@import "../Button/index";
@import "~antd/es/message/style/index";
@import "../../styles/variables";

.table-menu {
  margin-left: 5px;
  margin-bottom: 5px;
  &__button.ant-btn {
    border-radius: 50%;
  }
  &__icon {
    color: @primary-color;
    font-size: 14px !important;
    padding-top: 7px;
  }
}

.table-menu-dropdown {
  margin-right: 10px;
  &__icon {
    color: @primary-color;
    font-size: 14px !important;
    &__send {
      color: @primary-color;
      & > svg {
        transform: rotate(-40deg);
        margin-left: 2.7px;
        margin-bottom: 2.7px;
      }
    }
  }
  &__divider {
    margin: 4px 10px;
  }
}

.menu-item-checkbox {
  padding: unset;
  & > label.ant-checkbox-wrapper {
    padding: 5px 12px;
  }
}
