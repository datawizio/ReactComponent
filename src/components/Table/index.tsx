import clsx from "clsx";
import * as React from "react";
import { Table as AntdTable } from "antd";
import { reducer, initializer } from "./reducer";

import useColumns from "./hooks/useColumns";
import useDataSource from "./hooks/useDataSource";
import usePropsToState from "./hooks/usePropsToState";

import {
  useMemo,
  useReducer,
  useContext,
  useCallback,
  useImperativeHandle
} from "react";

import { TableContext } from "./context";
import ConfigContext from "../ConfigProvider/context";

import Loader from "../Loader";
import Cell from "./components/Cell";
import Column from "./components/Column";
import ToolBar from "./components/ToolBar";
import TableWrapper from "./components/TableWrapper";
import {
  LoadingOutlined,
  RightOutlined,
  DownOutlined
} from "@ant-design/icons";

import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";

import {
  basicSortHandler,
  basicFilterHandler,
  basicSearchHandler
} from "./utils/handlers";

import { TableProps, FCTable, TableRef } from "./types";

import "./index.less";
import useAsyncProviders from "./hooks/useAsyncProviders";

const Table = React.forwardRef<TableRef, TableProps>((props, ref) => {
  const {
    style,
    width,
    height,
    locale,
    isNested,
    showExpandIcon,
    children,
    components,
    dataProvider,
    showSizeChanger,
    dataProviderDeps,
    templatesProvider,
    rowChildrenProvider,
    nestedTableProvider,
    ...restProps
  } = props;

  const { translate } = useContext(ConfigContext);

  const [baseState, dispatch] = useReducer(reducer, props, initializer);

  const columnsState = useColumns(baseState, props);

  const dataSourceState = useDataSource(baseState, props);

  const state = {
    ...baseState,
    ...dataSourceState,
    ...columnsState,
    loading: false
  };

  const fetchData = useAsyncProviders(state, dispatch, props);

  usePropsToState(dispatch, props);

  const handleChangeTable = useCallback<TableProps["onChange"]>(
    (pagination, filters, sorter) => {
      dispatch({ type: "filter", payload: filters });
      dispatch({ type: "sort", payload: [].concat(sorter) });
      dispatch({ type: "paginate", payload: pagination as any });
    },
    []
  );

  const handleExpandRow = useCallback<TableProps["onExpand"]>(
    async (isExpanded, row) => {
      let toogle = true;
      if ((isNested && isNested(row)) || (!isNested && nestedTableProvider)) {
        if (!row.nested) {
          dispatch({
            type: "addLoadingRow",
            payload: row.key
          });
          const result = await nestedTableProvider(row);
          if (!result) toogle = false;
          dispatch({
            type: "setNestedTable",
            payload: [row, result]
          });
        }
      } else if (rowChildrenProvider && row.children && !row.children.length) {
        dispatch({
          type: "addLoadingRow",
          payload: row.key
        });
        const children = await rowChildrenProvider(row);
        dispatch({
          type: "setRowChildren",
          payload: [row, children.length ? children : undefined]
        });
      }

      toogle &&
        dispatch({
          type: isExpanded ? "expandRow" : "collapseRow",
          payload: row
        });
    },
    [rowChildrenProvider, nestedTableProvider, isNested]
  );

  const totalRenderer = useCallback(
    (total, [current, to]) => {
      return translate(locale.total, { current, to, total });
    },
    [translate, locale.total]
  );

  const expandIconRender = useCallback(
    ({ expanded, onExpand, record, prefixCls, expandable }) => {
      const iconPrefix = `${prefixCls}-row-expand-icon`;
      if (state.loadingRows[record.key]) return <LoadingOutlined />;
      let icon = null;
      let internalExpandable =
        expandable || (showExpandIcon && showExpandIcon(record));
      if (internalExpandable) {
        icon = <RightOutlined />;
        if (expanded) {
          icon = <DownOutlined />;
        }
      }

      return (
        <button
          type="button"
          onClick={e => {
            onExpand(record, e!);
            e.stopPropagation();
          }}
          className={clsx(iconPrefix, {
            [`${iconPrefix}-spaced`]: !internalExpandable,
            [`${iconPrefix}-expanded`]: expandable && expanded,
            [`${iconPrefix}-collapsed`]: expandable && !expanded
          })}
          aria-label={expanded ? locale.collapse : locale.expand}
        >
          {icon}
        </button>
      );
    },
    [state.loadingRows]
  );

  const customComponents = useMemo<TableProps["components"]>(
    () => ({
      ...components,
      table: props => <TableWrapper {...props} style={{ height, width }} />,
      header: {
        cell: props =>
          Boolean(props.model) ? <Column {...props} /> : <th {...props} />
      },
      body: {
        cell: props => <Cell {...props} />
      }
    }),
    [height, width, components]
  );

  const className = useMemo(
    () =>
      clsx(
        "dw-table",
        {
          "dw-table--loading": baseState.loading,
          "dw-table--empty": !state.dataSource.length,
          "dw-table--nestedable": props.expandable?.expandedRowRender
        },
        props.className
      ),
    [baseState.loading, props.className, state.dataSource.length]
  );

  useImperativeHandle(ref, () => ({
    async reload() {
      dispatch({ type: "loading", payload: true });
      await fetchData();
      dispatch({ type: "loading", payload: false });
    },
    updateRow(rowKey: string, data: any) {
      dispatch({ type: "updateRow", payload: [rowKey, data] });
    },
    addLoadingRow(rowKey: string) {
      dispatch({ type: "addLoadingRow", payload: rowKey });
    },
    resetPagination() {
      dispatch({ type: "resetPagination" });
    }
  }));

  return (
    <DndProvider backend={HTML5Backend}>
      <TableContext.Provider
        value={{
          tableProps: props,
          tableState: state,
          dispatch: dispatch,
          baseTableState: baseState
        }}
      >
        <Loader loading={Boolean(baseState.loading)}>
          {children}
          <AntdTable
            {...restProps}
            {...state}
            expandIcon={expandIconRender}
            className={className}
            onExpand={handleExpandRow}
            onChange={handleChangeTable}
            components={customComponents}
            pagination={
              props.pagination === false
                ? false
                : {
                    ...state.pagination,
                    ...props.pagination,
                    showTotal: totalRenderer
                  }
            }
          />
        </Loader>
      </TableContext.Provider>
    </DndProvider>
  );
}) as FCTable;

Table.defaultProps = {
  bordered: true,
  sortable: true,
  isResizableColumns: true,

  showSorterTooltip: false,

  className: "",
  size: "small",
  width: "auto",
  height: "auto",
  tableLayout: "fixed",

  locale: {
    total: "TABLE_TOTAL"
  },

  dTypesConfig: {},
  columnsConfig: {},
  cellRenderProps: {},

  columns: [],
  dataSource: [],
  pageSizeOptions: ["20", "35", "50", "100"],

  showSizeChanger: true,
  sortHandler: basicSortHandler,
  filterHandler: basicFilterHandler,
  searchHandler: basicSearchHandler
};

Table.ToolBar = ToolBar;

export default Table;
