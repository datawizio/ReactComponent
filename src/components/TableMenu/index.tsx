import {
  DownOutlined,
  SendOutlined,
  VerticalAlignBottomOutlined
} from "@ant-design/icons";
import { Dropdown, Menu, message } from "antd";
import React, { useCallback, useContext, useMemo } from "react";
import Button, { ButtonProps } from "../Button";
import Checkbox from "../Checkbox";
import ConfigContext from "../ConfigProvider/context";
import { TableContext } from "../Table/context";
import { TableState } from "../Table/types";
import { saveAs } from "file-saver";

import "./index.less";

export interface TableMenuProps extends ButtonProps {
  filename?: string;
  getFilename?: () => string;
  menuItems?: any;
  config?: any;
  duration?: number;
  exportHandler?: (
    tableState: TableState | null,
    filename: string
  ) => Promise<BlobPart> | Promise<void> | null;
  exportHandlerCallback?: (fileData: BlobPart | Blob, filename: string) => any;
  onSendClick?: () => Promise<void>;
  onTotalClick?: (e: any) => void;
  onExpandVertical?: (e: any) => void;
  onExpandHorizontal?: (e: any) => void;
}

const TableMenu: React.FC<TableMenuProps> = props => {
  const {
    config,
    onSendClick,
    exportHandler,
    exportHandlerCallback,
    getFilename,
    filename,
    duration,
    onTotalClick,
    onExpandHorizontal,
    onExpandVertical,
    ...restProps
  } = props;
  const { translate } = useContext(ConfigContext);

  const context = useContext(TableContext);

  const tableState = useMemo(() => {
    if (context) {
      return context.tableState;
    }
    return null;
  }, [context]);

  const handleExport = useCallback(
    async () => {
      if (exportHandler) {
        let file = filename;
        if (getFilename) {
          file = getFilename();
        }
        const messageKey = "exporting-" + file;

        message.loading({
          content: translate("LOADING"),
          key: messageKey,
          duration: duration
        });

        const fileData = await exportHandler(tableState, file);
        fileData && saveAs(new Blob([fileData]), file);

        message.success({ content: translate("SUCCESS"), key: messageKey });

        if (exportHandlerCallback && fileData) {
          await exportHandlerCallback(new Blob([fileData]), file);
        }
      }
    },
    //eslint-disable-next-line
    [
      tableState,
      translate,
      exportHandler,
      filename,
      getFilename,
      exportHandlerCallback
    ]
  );

  const handleMenuClick = e => {
    switch (e.key) {
      case "export_xlsx":
        handleExport();
        break;
      case "send_xlsx":
        onSendClick();
        break;
      default:
        break;
    }
  };
  const menu = (
    <Menu onClick={handleMenuClick}>
      {config?.fixed_total && (
        <Menu.Item key="fixed_total" className="menu-item-checkbox">
          <Checkbox onClick={onTotalClick}>{translate("FIXED_TOTAL")}</Checkbox>
        </Menu.Item>
      )}
      {config?.expand_table_vertically && (
        <Menu.Item key="expand_table_vertically" className="menu-item-checkbox">
          <Checkbox onClick={onExpandVertical}>
            {translate("EXPAND_THE_TABLE_VERTICALLY")}
          </Checkbox>
        </Menu.Item>
      )}
      {config?.expand_table_horizontally && (
        <Menu.Item
          key="expand_table_horizontally"
          className="menu-item-checkbox"
        >
          <Checkbox onClick={onExpandHorizontal}>
            {translate("EXPAND_TABLE_HORIZONTALLY")}
          </Checkbox>
        </Menu.Item>
      )}
      {(config?.fixed_total ||
        config?.expand_table_vertically ||
        config?.expand_table_horizontally) && (
        <Menu.Divider className={"table-menu__divider"} />
      )}
      <Menu.Item
        key="export_xlsx"
        icon={<VerticalAlignBottomOutlined className={"table-menu__icon"} />}
      >
        {translate("SAVE_XLS")}
      </Menu.Item>
      {config?.show_send_to_email && (
        <Menu.Item
          key="send_xlsx"
          icon={<SendOutlined className={"table-menu__icon__send"} />}
        >
          {translate("SEND_XLS")}
        </Menu.Item>
      )}
    </Menu>
  );
  return (
    <div className="table-menu table-toolbar--right">
      <Dropdown overlay={menu} trigger={["click"]}>
        <Button
          className="table-menu__button"
          icon={<DownOutlined className={"table-menu__icon"} />}
          border={false}
          {...restProps}
        />
      </Dropdown>
    </div>
  );
};

TableMenu.defaultProps = {
  filename: "exported_table.xlsx"
};

export default TableMenu;
