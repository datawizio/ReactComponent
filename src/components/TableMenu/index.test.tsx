import "jsdom-global/register";
import React from "react";
import { mount, shallow } from "enzyme";
import Table from "../Table";

import TableMenu from "./index";
import { render, fireEvent } from "@testing-library/react";

const dropdownClick = (wrapper, name) => {
  fireEvent.click(wrapper.getByText(name));
  return wrapper
    .getByText(name)
    .parentElement.querySelector(".ant-checkbox-input").checked;
};

const mockTableProps = {
  height: "auto",
  width: "auto",
  dTypesConfig: {},
  async: false,
  columns: [
    {
      title: "Name",
      key: "name",
      dataIndex: "name"
    }
  ],
  dataSource: [
    {
      key: "1",
      name: "John"
    }
  ]
};

const mockPropsTableMenu = {
  config: {
    fixed_total: true,
    expand_table_vertically: true,
    expand_table_horizontally: true,
    show_send_to_email: true
  }
};

const setUpWithoutTable = (props?) => mount(<TableMenu {...props} />);
const setUp = (props?) =>
  shallow(
    //@ts-ignore
    <Table {...mockTableProps}>
      <Table.ToolBar>
        <TableMenu config={props} />
      </Table.ToolBar>
    </Table>
  );

describe("TableMenu component", () => {
  let component;

  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn()
      }))
    });
  });

  beforeEach(() => {
    component = setUp().render();
  });
  it("should TableMenu render correctly", () => {
    expect(component).toMatchSnapshot();
  });
  it("should full TableMenu render correctly", () => {
    const wrapper = setUpWithoutTable(mockPropsTableMenu);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it.skip("should checkbox works correctly", async () => {
    const component = render(
      //@ts-ignore
      <Table {...mockTableProps}>
        <Table.ToolBar>
          <TableMenu
            {...{
              onTotalClick: jest.fn(),
              onExpandVertical: jest.fn(),
              onExpandHorizontal: jest.fn(),
              onSendClick: jest.fn(),
              exportHandler: jest.fn(),
              ...mockPropsTableMenu
            }}
          />
        </Table.ToolBar>
      </Table>
    );
    fireEvent.click(document.querySelector(".table-menu__button"));
    expect(dropdownClick(component, "FIXED_TOTAL")).toBeTruthy();
    expect(
      dropdownClick(component, "EXPAND_THE_TABLE_VERTICALLY")
    ).toBeTruthy();
    expect(dropdownClick(component, "EXPAND_TABLE_HORIZONTALLY")).toBeTruthy();
  });
  it.skip("should export button work correctly", () => {
    const wrapper = setUpWithoutTable({
      onTotalClick: jest.fn(),
      onExpandVertical: jest.fn(),
      onExpandHorizontal: jest.fn(),
      onSendClick: jest.fn(),
      exportHandler: jest.fn(),
      ...mockPropsTableMenu
    });
    wrapper.find(".table-menu__button").at(0).simulate("click");
    wrapper
      .findWhere(n => n.text() === "SEND_XLS")
      .first()
      .simulate("click");
    expect(wrapper.prop("onSendClick")).toBeCalled();
    wrapper
      .findWhere(n => n.text() === "SAVE_XLS")
      .first()
      .simulate("click");
    expect(wrapper.prop("exportHandler")).toBeCalled();
    expect(
      document.querySelectorAll(".ant-message-notice")[0].textContent
    ).toBe("LOADING");
  });
});
