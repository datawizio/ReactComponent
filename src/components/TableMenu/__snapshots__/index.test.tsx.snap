// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableMenu component should TableMenu render correctly 1`] = `
<div
  class="dw-table-container"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="dw-table__toolbar"
      >
        <div
          class="table-menu table-toolbar--right"
        >
          <button
            class="ant-btn ant-btn-icon-only dw-btn ant-dropdown-trigger table-menu__button no-border"
            type="button"
          >
            <span
              aria-label="down"
              class="anticon anticon-down table-menu__icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="ant-table-wrapper dw-table"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table ant-table-small ant-table-bordered ant-table-layout-fixed"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <div
                    class="dw-table__wrapper"
                    style="width:auto;height:auto"
                  >
                    <table>
                      <colgroup />
                      <thead
                        class="ant-table-thead"
                      >
                        <tr>
                          <th
                            class="dw-table__column dw-table__column--resizable ant-table-cell ant-table-column-has-sorters"
                            style="width:299px"
                            title="Name"
                          >
                            <div
                              class="ant-table-column-sorters"
                            >
                              <span>
                                Name
                              </span>
                              <span
                                class="ant-table-column-sorter ant-table-column-sorter-full"
                              >
                                <span
                                  class="ant-table-column-sorter-inner"
                                >
                                  <span
                                    aria-label="caret-up"
                                    class="anticon anticon-caret-up ant-table-column-sorter-up"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-up"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                  <span
                                    aria-label="caret-down"
                                    class="anticon anticon-caret-down ant-table-column-sorter-down"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="caret-down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="0 0 1024 1024"
                                      width="1em"
                                    >
                                      <path
                                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="ant-table-tbody"
                      >
                        <tr
                          class="ant-table-row ant-table-row-level-0"
                          data-row-key="1"
                        >
                          <td
                            class="ant-table-cell"
                          >
                            <div
                              class="dw-table__cell"
                            >
                              <div
                                class="dw-table__cell-data"
                              >
                                John
                              </div>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
            <ul
              class="ant-pagination mini ant-table-pagination ant-table-pagination-right"
              unselectable="unselectable"
            >
              <li
                class="ant-pagination-total-text"
              >
                TABLE_TOTAL
              </li>
              <li
                aria-disabled="true"
                class="ant-pagination-prev ant-pagination-disabled"
                title="Previous Page"
              >
                <button
                  class="ant-pagination-item-link"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    aria-label="left"
                    class="anticon anticon-left"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="left"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
              <li
                class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                tabindex="0"
                title="1"
              >
                <a
                  rel="nofollow"
                >
                  1
                </a>
              </li>
              <li
                aria-disabled="true"
                class="ant-pagination-next ant-pagination-disabled"
                title="Next Page"
              >
                <button
                  class="ant-pagination-item-link"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    aria-label="right"
                    class="anticon anticon-right"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="right"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
              <li
                class="ant-pagination-options"
              >
                <div
                  class="ant-select ant-select-sm ant-pagination-options-size-changer ant-select-single ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="undefined_list_0"
                        aria-autocomplete="list"
                        aria-controls="undefined_list"
                        aria-haspopup="listbox"
                        aria-owns="undefined_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        readonly=""
                        role="combobox"
                        style="opacity:0"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-item"
                      title="20 / page"
                    >
                      20 / page
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select:none;-webkit-user-select:none"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TableMenu component should full TableMenu render correctly 1`] = `
<div
  class="table-menu table-toolbar--right"
>
  <button
    class="ant-btn ant-btn-icon-only dw-btn ant-dropdown-trigger table-menu__button no-border"
    type="button"
  >
    <span
      aria-label="down"
      class="anticon anticon-down table-menu__icon"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </span>
  </button>
</div>
`;
